### PR TITLE
Mdk/33 cloudantquery

### LINKED ISSUE #33: Convert queries to new CloudantQuery syntax.
To get features such as compound conditions, queries on subentries, and multiple sort descriptors, we need to change the IS to use the new CloudantQuery support.

We currently use: https://github.com/jimix/CDTDatastore/blob/CDTIncrementalStore/doc/index-query.md

And we need to switch to: https://github.com/cloudant/CDTDatastore/blob/master/doc/query.md

### PR DESCRIPTION
**Pull request by mkistler**
Fixes #33

### CODE REVIEW

File: Classes/common/CDTIncrementalStore.m [hunk 1]
```diff
@@ -1767,59 +1639,73 @@ - (BOOL)loadMetadata:(NSError **)error
...
+- (NSArray *)sortForFetchRequest:(NSFetchRequest *)fetchRequest
 {
-    NSError *err = nil;
-    NSMutableDictionary *sdOpts = [NSMutableDictionary dictionary];
-
-    NSArray *sds = [fetchRequest sortDescriptors];
-    if (sds.count) {
-        if (sds.count > 1) {
-            oops(@"not sure what to do here");
+    NSMutableArray *keys = [NSMutableArray array];
```
Comment (mkistler) [inline Classes/common/CDTIncrementalStore.m]: This line no longer needed.
ACK

File: Classes/common/CDTIncrementalStore.m [hunk 2]
```diff
@@ -1767,59 +1639,73 @@ - (BOOL)loadMetadata:(NSError **)error
...
+    for (NSSortDescriptor *sd in [fetchRequest sortDescriptors]) {
+        NSString *sel = NSStringFromSelector([sd selector]);
+        if (![sel isEqualToString:@"compare:"]) {
+            [NSException raise:CDTISException format:@"we do not allow custom compares"];
         }
+        NSString *key = [sd key];
+        [keys addObject:key];
+        [sds addObject:@{ key : [sd ascending] ? @"asc" : @"desc" }];
+    }
+    return sds;
```
Comment (mkistler) [inline Classes/common/CDTIncrementalStore.m]: Convert back to NSArray.
ACK

File: Classes/common/CDTIncrementalStore.m [hunk 3]
```diff
@@ -1767,59 +1639,73 @@ - (BOOL)loadMetadata:(NSError **)error
...
+            [fields addObject:field];
+        }
     }
 
-    if ([sdOpts count]) {
-        return [NSDictionary dictionaryWithDictionary:sdOpts];
-    }
-    return nil;
+    NSString *indexName = [fields componentsJoinedByString:@"_"];
+    [self.datastore ensureIndexed:fields withName:indexName];
```
Comment (mkistler) [inline Classes/common/CDTIncrementalStore.m]: Check for success and log a failure with the logger.
ACK.

File: Classes/common/CDTIncrementalStore.m [hunk 4]
```diff
@@ -1859,14 +1744,23 @@ - (NSDictionary *)comparisonPredicate:(NSComparisonPredicate *)cp
     // process the predicate operator and create the key-value string
     NSPredicateOperatorType predType = [cp predicateOperatorType];
     switch (predType) {
+        case NSLessThanPredicateOperatorType:
+            result = @{ keyStr : @{@"$lt" : value} };
```
Comment (mkistler) [inline Classes/common/CDTIncrementalStore.m:1764]: Question: should we make constants for all the operators?

File: Classes/common/CDTIncrementalStore.m [hunk 5]
```diff
@@ -2086,57 +1970,53 @@ - (id)executeFetchRequest:(NSFetchRequest *)fetchRequest
...
-    // Get sort descriptors and add them as options
-    err = nil;
-    NSDictionary *options = [self processOptions:fetchRequest error:&err];
-    if (!options && err) {
-        if (error) *error = err;
-        // I think we do this on error, it is unclear
-        return nil;
-    }
-
     NSDictionary *query = [self queryForFetchRequest:fetchRequest error:&err];
```
Comment (mkistler) [inline Classes/common/CDTIncrementalStore.m]: What should I do with err here?  Get rid of it?
ACK - now create an error to return to the caller.

File: Tests/Tests/IncrementalStore.m [hunk 6]
```diff
@@ -420,7 +414,7 @@ - (void)testPredicates
     }
 
     /**
-     *  Fetch odd Numbers
+     *  Fetch boolean == value
      */
     NSPredicate *odd = [NSPredicate predicateWithFormat:@"check == NO"];
```
Comment (mkistler) [inline Tests/Tests/IncrementalStore.m:450]: Can we write a predicate that is just "check" (a boolean)?
NACK - no, there must be an operator.

**Comment by jimix** [comment]
Ack'ed

**Comment by jimix** [comment]
:disappointed: Replication Acceptance test for CoreData fails from the very first commit deb86b2 in `testCoreDataMigration`

**Comment by jimix** [comment]
Sorry, I could not resist peeking:
It seems that during the migration process uses a `TRUEPREDICATE`

``` diff
diff --git a/Classes/common/CDTIncrementalStore.m b/Classes/common/CDTIncrementalStore.m
index ed3a39c..b301bed 100644
--- a/Classes/common/CDTIncrementalStore.m
+++ b/Classes/common/CDTIncrementalStore.m
@@ -1880,8 +1880,8 @@ NSString *kNotOperator = @"$not";
     NSPredicate *fetchPredicate = [fetchRequest predicate];

     NSPredicate *fullPredicate;
-    if (fetchPredicate) {
-        // Avoid nested ANDs if possible
+    if (fetchPredicate && ![fetchPredicate isEqual:[NSPredicate predicateWithValue:YES]]) {
+    // Avoid nested ANDs if possible
         if ([fetchPredicate isKindOfClass:[NSCompoundPredicate class]] &&
             [((NSCompoundPredicate *)fetchPredicate)compoundPredicateType] == NSAndPredicateType) {
             fullPredicate = [NSCompoundPredicate andPredicateWithSubpredicates:[@[
```

You should amend/rebase/fixup this in deb86b2

**Comment by jimix** [comment]
All tests pass, I ACK this PR

**Pull request merged**